### ISSUE TITLE
Support for `sudo -e` or `sudoedit`

### ISSUE THREAD

**Issue by MrNossiom** [memorysafety/sudo-rs#762]
Hi 👋🏻,

## Description
`sudo` historically has a command to edit a file with the configured editor of your choice, often `sudo -e` or his own binary `sudoedit`. It copies the file with your user's permissions in a temporary place so that you can freely edit it.

The `-e` argument is already defined in the CLI options, so this could be more of a tracking issue.

Whether there should be a `sudoedit` binary distributed along the others can be discussed here. Since `sudo-rs` aims to be a fully functional replacement, including a `sudoedit` binary would help for backwards compatibility. It could also be an opportunity to drop a binary that mirrors the behavior of `sudo -e`

## Additional information
You can see an implementation of such a command here: [github.com/sudo-project/sudo:main/src/sudo_edit.c](https://github.com/sudo-project/sudo/blob/main/src/sudo_edit.c) (in C though 🤮)

I'm willing to try a PR for this 😄

**Comment by japaric** [comment]
Hello!

We discussed this topic in our last weekly meeting and we are happy to accept an external contribution that implements this feature.

We'd like to point out that there's a bit of a quality assurance / control process when landing new features that's not quite formalized (i.e. documented) but that we have been using:

- write [compliance tests](https://github.com/memorysafety/sudo-rs/tree/main/test-framework) for the new feature. I personally like to take a bit of a TDD (Test Driven Development) approach and write the tests first and then write enough code to make the tests pass; then repeat as necessary

- check [original sudo's CVE list](https://www.sudo.ws/security/advisories/) to see if there's anything related to the feature. it's good to keep those in mind when writing the code; sometimes the CVE inform about edge cases that may not be obvious

Our CI should catch minor issues like clippy warnings and rustfmt but it's good to run them locally before opening a PR

**Comment by squell** [comment]
Hi! Any updates on this issue? 😃

**Comment by mrnossiom** [comment]
Hi, I'd love to contribute, but I cannot find much energy for this right now. I'd be happy to review or contribute later. I really don't want to block anything, and any progress I'll do we'll be mentioned here. 😃

**Comment by squell** [comment]
Having had a discussion internally, we came up with this roadmap; some of these tasks might have slight overlap, but a lot could also happen in parallel.

- [x] in CLI: add logic for `sudo -e` and route that to a (stub) separate sudoedit pipeline
- [x] in CLI: add logic so that if sudo is executed as `sudoedit`, that routes to the same pipeline as `sudo -e` (i.e. the mode of operation is that sudoedit will become a symlink to sudo). We should also make sure that this part if easily modded by downstream packagers in case they would want to ship something as cumbersome as `sudoedit-rs`. (https://github.com/trifectatechfoundation/sudo-rs/pull/1195)
- [x] write compliance tests for `sudo -e` using a dummy editor
- [x] in sudoers/AST: add recognition of the "sudoedit" pseudocommand (similar to how this was recently done for "list"; it's fine to have the slightly "stringly typed" for now); there's already a stub for this. (#1152) 
- [x] probably in sudoers/policy: add a high level check to see if a user is allowed to edit a file; this should translate to a regular permission check for the "sudoedit" pseudocommand
- [x] draft the aforementioned sudoedit pipeline similar to how the su and sudo pipelines work, this would have subtasks (https://github.com/trifectatechfoundation/sudo-rs/pull/1199):
  - [x] create logic for copying a file back-and-forth in the privileged context (can be copied/inspired by from visudo, code sharing is a "nice to have" initially, but the security requirements for visudo and sudoedit might be different)
  - [x] spawn an editor in the same privilege context as the calling user
- [x] take a new look at the sudoedit-related CVE's of ogsudo and make sure we avoid them; also study ogsudo code a bit for bits of safety it has; update https://github.com/trifectatechfoundation/sudo-rs/blob/main/docs/sudo-cve.md
- [x] look for `Defaults` that might be useful that modify `sudoedit` behaviour, and that fit our Berlin Criteria

**Comment by squell** [comment]
@mrnossiom Since you already did some work on the CLI side of things, do you think you have the time to work your PR to conform to the first two bullet points? (The sudoedit pipeline itself can be a simple `todo!()`); otherwise we can pick that up based off of your draft.

Of course compliance tests (there are several examples for visudo that could serve as inspiration) are also welcome.

**Comment by bjorn3** [comment]
In the logic for copying back-and-forth og sudo has had several CVEs:

* https://www.sudo.ws/security/advisories/sudoedit/ where replacing the copy with a symlink would allow getting the contents of arbitrary files.
* https://bugzilla.sudo.ws/show_bug.cgi?id=707 where if the file to edit is owned by the user, they can replace it with a symlink to allow overwriting arbitrary files. This was fixed by introducing the FOLLOW/NOFOLLOW tags.
* https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-23239 which allowed testing for the existence of directories through a race condition. I believe this was fixed in https://github.com/sudo-project/sudo/commit/db1f27c0350e9e437c93780ffe88648ae1984467

We need to be very careful about where to allow symlinks. In particular the copy needs to be either opened using O_NOFOLLOW or maybe opened by a helper process running with the same privileges as the user that runs sudo. And for the file to be edited, we either need to respect FOLLOW/NOFOLLOW or enforce NOFOLLOW. The default since sudo 1.8.15 is NOFOLLOW as `sudoedit_follow` is not enabled by default. I personally think we should not implement `sudoedit_follow`, but should allow using the `FOLLOW` tag.

Edit: These are notes to myself for implementing this logic, but someone double checking my investigation can't hurt.

**Comment by mrnossiom** [comment]
I'll start to work on this PR Saturday morning. I'll conform to the bullet points as you mentioned.

EDIT: I rebased and continued locally. I am leaving until June 29th or so. I'll then have plenty of time as I've finished all my exams.

**Comment by squell** [comment]
> In the logic for copying back-and-forth og sudo has had several CVEs:
> 
> * https://www.sudo.ws/security/advisories/sudoedit/ where replacing the copy with a symlink would allow getting the contents of arbitrary files.
> * https://bugzilla.sudo.ws/show_bug.cgi?id=707 where if the file to edit is owned by the user, they can replace it with a symlink to allow overwriting arbitrary files. This was fixed by introducing the FOLLOW/NOFOLLOW tags.
> * https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-23239 which allowed testing for the existence of directories through a race condition. I believe this was fixed in [sudo-project/sudo@db1f27c](https://github.com/sudo-project/sudo/commit/db1f27c0350e9e437c93780ffe88648ae1984467)

👍 

> We need to be very careful about where to allow symlinks. In particular the copy needs to be either opened using O_NOFOLLOW or maybe opened by a helper process running with the same privileges as the user that runs sudo. And for the file to be edited, we either need to respect FOLLOW/NOFOLLOW or enforce NOFOLLOW. The default since sudo 1.8.15 is NOFOLLOW as `sudoedit_follow` is not enabled by default. I personally think we should not implement `sudoedit_follow`, but should allow using the `FOLLOW` tag.
> 
> Edit: These are notes to myself for implementing this logic, but someone double checking my investigation can't hurt.

I'm a bit stricter, I think implementing`FOLLOW:` is a separate feature that we must track in a different issue; that allows focussing on its necessity. For the first release of sudoedit we don't need it. The default should be 'NOFOLLOW'.

**Comment by squell** [comment]
Relevant `Defaults`:

* `sudoedit_follow`: as discussed, we keep this `off` (the default for ogsudo)
* `sudoedit_checkdir`: we discussed this offline and we probably want to keep this `on` (the default for ogsudo); although the attack it mitigates against appears less applicable to our sudoedit given the way we organize things (it's more a prevention against misconfiguration)
* `editor`: tracked in #1183
